### PR TITLE
DEPR: deprecate .get_value and .set_value for Series, DataFrame, Panel & Sparse

### LINKED ISSUE #15269: DEPR: set_value / get_value
we already have too many public indexers.....

### PR DESCRIPTION
**Pull request by jreback**
closes #15269

### CODE REVIEW

**Comment by jreback** [comment]
@TomAugspurger @jorisvandenbossche

File: pandas/core/sparse/frame.py [hunk 1]
```diff
@@ -460,7 +497,35 @@ def set_value(self, index, col, value, takeable=False):
         -------
         frame : DataFrame
         """
-        dense = self.to_dense().set_value(index, col, value, takeable=takeable)
+        warnings.warn("get_value is deprecated and will be removed "
```
Comment (jschendel) [inline pandas/core/sparse/frame.py]: Should this one be "set_value is deprecated..." instead of "get_value"?

Comment (jreback) [reply pandas/core/sparse/frame.py]: thanks...i fixed the whole doc-string duplication mess as well.

**Comment by jorisvandenbossche** [comment]
I am not fully convinced this is needed. As @wesm said in the issue, they were originally meant as fast access methods, and although performance has degraded over time, they are still considerably faster as any alternative.

I agree in the large majority of the use case you should not (or can avoid) having to set / get individual values, but why not keep them for those few cases you cannot otherwise?

**Comment by jreback** [comment]
well we have more than 1 way to do things (12 ways in fact). This is just a bad API design. Not to mention that these barely have any tests / attention, and they break in lots of cases (as they don't handle validation / dtypes propertly).

this is a no-brainer.

**Comment by jreback** [comment]
``.iat/.at`` are properly supported and pretty close in perf

**Comment by jreback** [comment]
any final comments?

File: doc/source/whatsnew/v0.21.0.txt [hunk 2]
```diff
@@ -666,6 +666,7 @@ Deprecations
 - ``cdate_range`` has been deprecated in favor of :func:`bdate_range`, which has gained ``weekmask`` and ``holidays`` parameters for building custom frequency date ranges. See the :ref:`documentation <timeseries.custom-freq-ranges>` for more details (:issue:`17596`)
 - passing ``categories`` or ``ordered`` kwargs to :func:`Series.astype` is deprecated, in favor of passing a :ref:`CategoricalDtype <whatsnew_0210.enhancements.categorical_dtype>` (:issue:`17636`)
 - Passing a non-existant column in ``.to_excel(..., columns=)`` is deprecated and will raise a ``KeyError`` in the future (:issue:`17295`)
+- ``.get_value`` and ``.set_value`` on ``Series``, ``DataFrame``, ``Panel``, ``SparseSeries``, and ``SparseDataFrame`` are deprecated in favor of using ``.iat[]`` or ``.at[]`` accessors
```
Comment (jschendel) [inline doc/source/whatsnew/v0.21.0.txt]: Should there be an issue number here?

Unrelated, but if you're going to make a change, there's also a typo on the line above this one that could be fixed: "non-existant" -> "non-existent".  Can submit a fix myself if there's no need to add an issue number and changes don't need to be made.

Comment (jreback) [reply doc/source/whatsnew/v0.21.0.txt]: thanks, both fixed

**Comment by jorisvandenbossche** [comment]
I already stated my reservations about the need for this. 
So ideally it would be good to hear some more opinions from other devs (because if most are OK with this, it's fine for me too, I don't care about it *too* much)

**Comment by TomAugspurger** [comment]
I feel the same as Joris, and will defer to others on this.

**Comment by wesm** [comment]
As we have already capitulated long ago on caring about microperformance (https://mail.python.org/pipermail/pandas-dev/2016-March/000499.html) I don't have strong feelings. If it does not cause us a maintenance hardship (maybe it does...), and the performance is better, why not keep them?

**Comment by jreback** [comment]
perf is almost exactly the same as ``.iat/.at``; they are in fact wrappers around the internal versions of ``.set_value/.get_value``. but absolutely this is an API/maintenance burden. too much surface area.

**Comment by jorisvandenbossche** [comment]
Since they are indeed rather thin wrappers around the internal set/get_value, I don't think they are actually that much of a maintenance burden. But it is true it *is* an API 'burden' (they add to the long list of available functions that users have to cope with). 
So it's a bit a trade-off. But they are still a factor of around 3x faster as `.at`. So given that I would tend to keep them (although I never use them myself).

**Comment by jreback** [comment]
you are completely missing the point.

They duplicate API. They are not tested in the least and they are very buggy. The entire point of .iat/.at is that they follow a common path and ARE tested with all dtypes. 

We are not going to keep around buggy code that has duplicate API's. This such a maintance burden. We need to actively be removing code, not preserving untested, non-documented duplicative API's.

**Comment by wesm** [comment]
I wouldn't say he / I are missing the point but we have a slightly different perspective =) I am +0 with removing the APIs in the interest of simplification and maintainability.

**Comment by jorisvandenbossche** [comment]
If you read my previous post, I think you can see that I am not completely missing the point. I *agree* that it is duplicate API. But there is still a trade-off between duplicate api vs speed / breaking people's code. 
But I do not agree they are not tested. They are tested by the many indexing tests that use them under the hood.

But to the point: by further looking at it, I am certainly fine with the idea of deprecating those functions. As I would also like to clean up the API, and iat/at are I think nicer. 
But, `at`/`iat` are actually also only very small wrappers around `get_value`/`set_value`. They also do almost as little data validation as get_value/set_value (eg the bug report about PeriodIndex with get_value (https://github.com/pandas-dev/pandas/issues/15268), you have exactly the same using `at`). And I think in general this limited validation is actually the feature of those functions/methods (although the fact it cannot handle a Period on a PeriodIndex could be seen as a bug for `at`). 
And the little bit extra validation they do (the reason they are slower) is IMO not that useful. So I would like to hear if you would be OK with making at/iat a bit faster (closer to get_value/set_value).

Eg, if you look at the `at` implementation for getting, it does the following validation:

- check that it is a tuple -> otherwise raise ValueError
- on integer axis, check that key is integer -> otherwise ValueError
- on non-integer axis, check that key is not integer -> otherwise raise ValueError

But in all those case, when not doing this check, you would get a KeyError otherwise with the `get_value` that is used under the hood. But those checks make up for a big part of the reason that `at` is 3x as slow as `get_value`. 
Are those checks necessary? They provide a little bit a more informative error message (compared to just KeyError). And dropping the checks would be an API change from ValueError to KeyError. But IMO they are not really important, and I would be in favor to drop them, making `at` closer in speed to `get_value`.

Would you be with such a change?

(and to repeat from above, I am OK with the idea of removing `get_value`/`set_value`)

**Review by jorisvandenbossche — COMMENTED** [review]
Minor doc comment, for the rest no code comments!

**Comment by jreback** [comment]
> Eg, if you look at the at implementation for getting, it does the following validation:

> check that it is a tuple -> otherwise raise ValueError
on integer axis, check that key is integer -> otherwise ValueError
on non-integer axis, check that key is not integer -> otherwise raise ValueError
But in all those case, when not doing this check, you would get a KeyError otherwise with the get_value that is used under the hood. But those checks make up for a big part of the reason that at is 3x as slow as get_value.
Are those checks necessary? They provide a little bit a more informative error message (compared to just KeyError). And dropping the checks would be an API change from ValueError to KeyError. But IMO they are not really important, and I would be in favor to drop them, making at closer in speed to get_value.

> Would you be with such a change?

I guess my overarching point is that we have a framework that we can/should test ``.iat/at`` more. So for sure testing is +1. We already have some scalar fast paths that least to these routines, but some of those changes are certainly possible/welcome. Obviously not averse to speeding things up.

Things done in the unified indexing context are always welcome.

didn't see any doc comments left, can you point.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -1918,6 +1919,8 @@ def get_value(self, index, col, takeable=False):
         """
         Quickly retrieve single value at passed column and index
 
+        .. deprecated:: 0.21.0
+
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: Can you add here the alternative like in the deprecations message? (the "use '.at[]' or '.iat[]' instead")

(and the same for the other docstrings)

**Review by jorisvandenbossche — COMMENTED** [review]
Oops, forgot to push the button. Here is the minor doc comment.

**Pull request merged**